File: app/src/main/java/ru/santaev/clipboardtranslator/model/repository/LanguageRepository.java
package ru.santaev.clipboardtranslator.model.repository;

import android.arch.lifecycle.LiveData;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import ru.santaev.clipboardtranslator.api.IApiService;
import ru.santaev.clipboardtranslator.api.LanguagesResponse;
import ru.santaev.clipboardtranslator.db.AppDatabase;
import ru.santaev.clipboardtranslator.db.dao.LanguageDao;
import ru.santaev.clipboardtranslator.db.entity.Language;

public class LanguageRepository {

    private IApiService apiService;
    private LanguageDao languageDao;

    private LiveData<List<Language>> languages;

    public LanguageRepository(IApiService apiService) {
        this.apiService = apiService;
        languageDao = AppDatabase.getInstance().getLanguageDao();
    }

    public LiveData<List<Language>> getLanguages() {
        loadLanguages();
        if (languages == null) {
            languages = languageDao.getLanguages();
        }
        return languages;
    }

    private void loadLanguages(){
        Single.fromCallable(() -> apiService.getLanguages())
                .subscribeOn(Schedulers.io())
                .subscribe(this::onLoadLanguages, Throwable::printStackTrace);
    }

    private void onLoadLanguages(LanguagesResponse languagesResponse) {
        Map<String, String> languagesRaw = languagesResponse.getLanguages();
        ArrayList<Language> languages = new ArrayList<>();
        for (String code : languagesRaw.keySet()) {
            languages.add(new Language(0, code, languagesRaw.get(code)));
        }
        languageDao.insertAll(languages);
    }

}

File: app/src/main/java/ru/santaev/clipboardtranslator/model/repository/LanguageRepository.kt
package ru.santaev.clipboardtranslator.model.repository

import io.reactivex.Flowable
import io.reactivex.schedulers.Schedulers
import ru.santaev.clipboardtranslator.api.IApiService
import ru.santaev.clipboardtranslator.api.LanguagesResponse
import ru.santaev.clipboardtranslator.db.dao.ILanguageDao
import ru.santaev.clipboardtranslator.db.entity.Language

class LanguageRepository(
        private val languageDao: ILanguageDao,
        private val apiService: IApiService
) : ILanguageRepository {

    private var languages: Flowable<List<Language>>? = null

    override fun getLanguages(): Flowable<List<Language>> {
        loadLanguages()

        languages.let {
            return if (it == null) {
                val newLanguages: Flowable<List<Language>> = languageDao.languages
                languages = newLanguages
                return newLanguages
            } else {
                it
            }
        }
    }

    private fun loadLanguages() {
        apiService.getLanguages()
                .subscribeOn(Schedulers.io())
                .subscribe(this::onLoadLanguages, { it.printStackTrace() })
    }

    private fun onLoadLanguages(languagesResponse: LanguagesResponse) {
        val languagesRaw = languagesResponse.languages
        val languages = languagesRaw?.keys?.map { Language(0, it, languagesRaw[it]) }
        languageDao.insertAll(languages)
    }

}
